### PR TITLE
refactor(animations): add an onStart handler for AnimationPlayer

### LINKED ISSUE #10304: Animation "onDone" callback support missing
We need this:

```html
<div [@animation]="exp" (@animation.done)="onDone()" (@animation.start)="onStart()">...</div>
```

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

**Comment by matsko** [comment]
@ThomasBurleson @robertmesserle

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 1]
```diff
@@ -385,60 +391,92 @@ function _getStylesArray(obj: any): {[key: string]: any}[] {
...
+    });
+
+    outputAsts.forEach(output => {
+      if (output.name[0] == '@') {
+        var normalizedOutputData = parseAnimationOutputName(output.name.substr(1), this.errors);
+        let triggerName = normalizedOutputData.name;
+        let triggerEventPhase = normalizedOutputData.phase;
+        if (!animationRegistry[triggerName]) {
+          this.errors.push(new AnimationParseError(
+              `couldn't find the corresponding ${isHostLevel ? 'host-level ' : '' }animation trigger definition for (@${triggerName})`));
```
Comment (IgorMinar) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: capital C?

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 2]
```diff
@@ -385,60 +391,92 @@ function _getStylesArray(obj: any): {[key: string]: any}[] {
...
+      animationRegistry: {[key: string]: any}, isHostLevel: boolean): void {
+    var detectedAnimationInputs: {[key: string]: boolean} = {};
+    inputAsts.forEach(input => {
+      if (input.type == t.PropertyBindingType.Animation) {
+        var triggerName = input.name;
+        if (isPresent(animationRegistry[triggerName])) {
+          detectedAnimationInputs[triggerName] = true;
+        } else {
+          this.errors.push(
+              new AnimationParseError(`couldn't find an animation entry for ${triggerName}`));
```
Comment (IgorMinar) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: capital C

File: modules/@angular/platform-browser/src/dom/web_animations_player.ts [hunk 3]
```diff
@@ -66,10 +67,17 @@ export class WebAnimationsPlayer implements AnimationPlayer {
...
 
-  onDone(fn: Function): void { this._subscriptions.push(fn); }
+  onStart(fn: () => void): void { this._onStartFns.push(fn); }
+
+  onDone(fn: () => void): void { this._onDoneFns.push(fn); }
 
   play(): void {
     this.init();
+    if (!this.hasStarted()) {
+      this._onStartFns.forEach(subscription => subscription());
```
Comment (IgorMinar) [inline modules/@angular/platform-browser/src/dom/web_animations_player.ts]: rename subscription to fn?

File: modules/@angular/core/testing/mock_animation_player.ts [hunk 4]
```diff
@@ -34,20 +35,26 @@ export class MockAnimationPlayer implements AnimationPlayer {
...
+  onDone(fn: () => void): void { this._onDoneFns.push(fn); }
+
+  onStart(fn: () => void): void { this._onStartFns.push(fn); }
 
   hasStarted() { return this._started; }
 
   play(): void {
-    this._started = true;
+    if (!this.hasStarted()) {
+      this._onStartFns.forEach((entry) => entry());
```
Comment (IgorMinar) [inline modules/@angular/core/testing/mock_animation_player.ts]: entry -> fn

File: modules/@angular/core/src/animation/animation_sequence_player.ts [hunk 5]
```diff
@@ -50,22 +51,28 @@ export class AnimationSequencePlayer implements AnimationPlayer {
       if (!isPresent(this.parentPlayer)) {
         this.destroy();
       }
-      this._subscriptions.forEach(subscription => subscription());
-      this._subscriptions = [];
+      this._onDoneFns.forEach(subscription => subscription());
```
Comment (IgorMinar) [inline modules/@angular/core/src/animation/animation_sequence_player.ts]: fn

File: modules/@angular/core/src/animation/animation_player.ts [hunk 6]
```diff
@@ -32,19 +33,27 @@ export abstract class AnimationPlayer {
...
+  private _onDoneFns: Function[] = [];
+  private _onStartFns: Function[] = [];
   private _started = false;
   public parentPlayer: AnimationPlayer = null;
   constructor() { scheduleMicroTask(() => this._onFinish()); }
   /** @internal */
   _onFinish() {
-    this._subscriptions.forEach(entry => { entry(); });
-    this._subscriptions = [];
+    this._onDoneFns.forEach(entry => entry());
```
Comment (IgorMinar) [inline modules/@angular/core/src/animation/animation_player.ts]: fn

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 7]
```diff
@@ -1005,6 +974,312 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+             expect(loadingCalls).toEqual(2);
+           })();
+         }));
+    });
+
+    describe('ng directives', () => {
+      describe('[ngClass]', () => {
+        it('should persist ngClass class values when a remove element animation is active',
+           inject(
+               [TestComponentBuilder, AnimationDriver],
```
Comment (IgorMinar) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: remove TCB

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 8]
```diff
@@ -1005,6 +974,312 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+               fakeAsync(
+                   (tcb: TestComponentBuilder, driver: InnerContentTrackingAnimationDriver) => {
+                     makeAnimationCmp(
+                         tcb, `<div [ngClass]="exp2" *ngIf="exp" @trigger></div>`,
+                         [
+                           trigger('trigger', [transition('* => void', [animate(1000)])]),
+                         ],
+                         (fixture: any /** TODO #9100 */) => {
```
Comment (IgorMinar) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: add type ComponentFixture

**Comment by IgorMinar** [comment]
lgtm but @mhevery should take a look as well since this change is non-trivial and I'm not familiar with parts of the codebase it touches

**Comment by mhevery** [comment]
Please take care of igor's comments, LGTM.

**Comment by matsko** [comment]
Closed in favour of https://github.com/angular/angular/pull/10892

**Comment by IgorMinar** [comment]
please reformat

**Pull request merged**